src/pybind/mgr/prometheus/module.py

```diff
@@ -1292,9 +1299,13 @@ def respond():
         # wait for the shutdown event
         self.shutdown_event.wait()
         self.shutdown_event.clear()
+        # tell metrics collection thread to stop collecting new metrics
+        self.metrics_thread.stop()
         cherrypy.engine.stop()
         self.log.info('Engine stopped.')
         self.shutdown_rbd_stats()
+        # wait for the metrics collection thread to stop
+        self.metrics_thread.join()
```

Comment:
So it will be waiting up to `scrape_interval` second until it wakes up from sleep. It could be improved by replacing the `sleep(scrape_interval)` with `condition.wait(scrape_interval)` and calling `condition.notify()` here before join.

Reply:
Yeah, it is not ideal but it doesn't seem to be blocking either (in my testing). The mgr daemon is restarted quickly even with this line in it. We could change it to use condition.wait instead though.

Reply:
Interesting, what value did you have for `scrape_interval` in your tests? The default used to be 5 sec but (recently?) has been changed to 15 sec, and I would expect the shutdown delay `scrape_interval / 2` sec average due to waiting for the sleep to finish.

Reply:
The `scrape_interval` was set to 15 seconds. The kill might be aggressive here though.

Anyway, I used `threading.Event` to improve this. We don't need to use `threading.Condition` as we do not need any lock here.